sequencer/src/api/sql.rs

```diff
@@ -344,26 +351,68 @@ async fn reconstruct_state<'a>(
         "loaded fee state does not match parent header"
     );
 
-    // For catchup that is required during the following state replay, use the local database via
-    // this transaction.
-    let peers = SqlStateCatchup::new(Arc::new(tx), Default::default());
+    // We need the blocks frontier as well, to apply the STF.
+    let frontier = load_frontier(tx, from_height)
+        .await
+        .context("unable to reconstruct state because frontier is not available at origin")?;
+    match frontier
+        .proof
+        .first()
+        .context("empty proof for frontier at origin")?
+    {
+        MerkleNode::Leaf { pos, elem, .. } => state
+            .block_merkle_tree
+            .remember(*pos, *elem, frontier)
+            .context("failed to remember frontier")?,
+        _ => bail!("invalid frontier proof"),
+    }
 
     // Apply subsequent headers to compute the later state.
     for proposal in &leaves {
-        state = compute_state_update(&state, instance, &peers, parent, proposal)
-            .await
-            .context(format!(
-                "unable to reconstruct state because state update {} failed",
-                proposal.height(),
-            ))?
-            .0;
+        state = compute_state_update(
+            &state,
+            instance,
+            &NullStateCatchup::default(),
+            parent,
+            proposal,
+        )
+        .await
+        .context(format!(
+            "unable to reconstruct state because state update {} failed",
+            proposal.height(),
+        ))?
+        .0;
         parent = proposal;
     }
 
     tracing::info!(from_height, ?to_view, "successfully reconstructed state");
     Ok((state, to_leaf))
 }
 
+/// Get the set of all fee accounts on which the STF depends, applied to the given list of headers.
+async fn fee_account_dependencies(
+    instance: &NodeState,
+    mut parent: &Leaf,
```

Comment:
I don't think this only lets you reassign parent to a different reference in this function but wanted to double check

Reply:
That's the idea, I don't want to modify `parent` in the caller because we don't want to lose track of where catchup is starting from, but we need `parent` to advance in the callee. Effectively this function is doing like a dry run of the whole replay process, not actually executing but only looking for dependencies